### PR TITLE
do not cache output if page is redirected

### LINKED ISSUE #4278: Profile controls loading for unauthenticated users when loading via public page with MemoryOutputCaching enabled.
## Description of bug
Profile controls loading for unauthenticated users when called on top of public page with MemoryOutputCaching enabled.

## Steps to reproduce
1. Login as host user
2. Go to Persona Bar -> Content -> Pages 
3. Click "Add Page" button and give a name e.g., test
4. Click on "Permissions" tab and check "View" permissions for "All Users"
5. Click on "Advanced" tab and then click on "More" sub-tab
6. Choose "MemoryOutputCachingProvider" option for "Output Cache Provider"
7. Provide "86400" for "Cache Duration (seconds)"
8. Click "Add Page" button to create the page.
9. Logout
10. Browse to domain.com/test/ctl/profile and it will show a login tab
11. Browse again to domain.com/ctl/profile

## Current behavior
Shows "Object moved to here" and also show profile controls even to unauthenticated users.

## Expected behavior
It should not show profile controls to unauthenticated users.

## Screenshots
image

## Error information
Provide any error information (console errors, error logs, etc.) related to this bug.

## Additional context
Provide any additional context that may be helpful in understanding and/or resolving the bug.

## Affected version

* [x] 10.00.00 alpha build
* [x] 09.08.00 release candidate
* [x] 09.07.02 latest supported release

## Affected browser

* [x] Chrome
* [x] Firefox
* [x] Safari
* [x] Internet Explorer 11
* [x] Microsoft Edge (Classic)
* [x] Microsoft Edge Chromium

### PR DESCRIPTION
**Pull request by tauqeer-haider**
Fixes #4278 

## Summary

Prevented caching of the page output if the page is redirected.

### CODE REVIEW

**Review by bdukes — CHANGES_REQUESTED** [review]
Thanks for this!  I think I would prefer this check to be a level higher, in [`OutputCacheModule`](https://github.com/dnnsoftware/Dnn.Platform/blob/a0793aab185567505c7d7b70548b5e5596d69601/DNN%20Platform/HttpModules/OutputCaching/OutputCacheModule.cs#L247-L257), which already has references to the context and response, and already does a check on user agent.

File: DNN Platform/HttpModules/OutputCaching/OutputCacheModule.cs [hunk 1]
```diff
@@ -249,7 +250,8 @@ private void OnUpdateRequestCache(object sender, EventArgs e)
             if (!HttpContext.Current.Request.Browser.Crawler)
```
Comment (bdukes) [inline DNN Platform/HttpModules/OutputCaching/OutputCacheModule.cs:250]:
```suggestion
            var request = HttpContext.CurrentRequest;
            var isRedirect = response.StatusCode == (int)HttpStatusCode.Redirect;
            if (!request.Browser.Crawler && !isRedirect)
```

File: DNN Platform/HttpModules/OutputCaching/OutputCacheModule.cs [hunk 2]
```diff
@@ -249,7 +250,8 @@ private void OnUpdateRequestCache(object sender, EventArgs e)
...
             {
                 var responseFilter = this._app.Context.Items[ContextKeyResponseFilter] as OutputCacheResponseFilter;
-                if (responseFilter != null)
+                if (responseFilter != null &&
+                    HttpContext.Current.Response.StatusCode != Convert.ToInt32(HttpStatusCode.Redirect))
```
Comment (bdukes) [inline DNN Platform/HttpModules/OutputCaching/OutputCacheModule.cs:254]:
```suggestion
                if (responseFilter != null)
```

**Review by bdukes — APPROVED** [review]
Thanks!  Let's move the check up higher, but otherwise this looks good to me 👍🏻

File: DNN Platform/HttpModules/OutputCaching/OutputCacheModule.cs [hunk 3]
```diff
@@ -2,258 +2,261 @@
...
+                        duration);
+                    this._app.Context.Items[ContextKeyResponseFilter] = responseFilter;
+                    this._app.Context.Response.Filter = responseFilter;
+                }
+            }
+        }
+
+        private void OnUpdateRequestCache(object sender, EventArgs e)
+        {
+            var request = HttpContext.CurrentRequest;
```
Comment (bdukes) [inline DNN Platform/HttpModules/OutputCaching/OutputCacheModule.cs:250]:
```suggestion
            var request = HttpContext.Current.Request;
```

**Review by valadas — APPROVED** [review]
Looks good to me, thanks.

**Comment by valadas** [comment]
Oops, there is this build error though:

```
"D:\a\1\s\DNN_Platform.sln" (Rebuild target) (1) ->
       "D:\a\1\s\DNN Platform\DotNetNuke.Web\DotNetNuke.Web.csproj.metaproj" (Rebuild target) (6) ->
       "D:\a\1\s\DNN Platform\HttpModules\DotNetNuke.HttpModules.csproj" (Rebuild target) (19) ->
       (CoreCompile target) -> 
         OutputCaching\OutputCacheModule.cs(251,30): error CS0103: The name 'response' does not exist in the current context [D:\a\1\s\DNN Platform\HttpModules\DotNetNuke.HttpModules.csproj]
```

File: DNN Platform/HttpModules/OutputCaching/OutputCacheModule.cs [hunk 4]
```diff
@@ -2,258 +2,261 @@
...
+                    this._app.Context.Items[ContextKeyResponseFilter] = responseFilter;
+                    this._app.Context.Response.Filter = responseFilter;
+                }
+            }
+        }
+
+        private void OnUpdateRequestCache(object sender, EventArgs e)
+        {
+            var request = HttpContext.Current.Request;
+            var isRedirect = response.StatusCode == (int)HttpStatusCode.Redirect;
```
Comment (valadas) [inline DNN Platform/HttpModules/OutputCaching/OutputCacheModule.cs:251]: response does not exist here (build failed).

**Review by valadas — CHANGES_REQUESTED** [review]

**Review by valadas — APPROVED** [review]
Awesome, thanks

**Pull request merged**